### PR TITLE
CPP-188 - Added performance and dianostic metrics

### PR DESCRIPTION
**Pull request by mpenick**

### CODE REVIEW

File: include/cassandra.h [hunk 1]
```diff
@@ -305,6 +305,44 @@ typedef struct CassSchemaMetaField_ CassSchemaMetaField;
...
+  struct {
+    cass_uint64_t min; /**< Minimum in microseconds */
+    cass_uint64_t max; /**< Maximum in microseconds */
+    cass_uint64_t mean; /**< Mean in microseconds */
+    cass_uint64_t stddev; /**< Standard deviation in microseconds */
+    cass_uint64_t median; /**< Median in microseconds */
+    cass_uint64_t percentile_75th; /**< 75th percentile in microseconds */
+    cass_uint64_t percentile_95th; /**< 95th percentile in microseconds */
+    cass_uint64_t percentile_98th; /**< 98th percentile in microseconds */
+    cass_uint64_t percentile_99th; /**< 99the percentile microseconds */
```
Comment (aholmberg) [inline include/cassandra.h]: small typo

File: src/control_connection.cpp [hunk 2]
```diff
@@ -175,6 +175,7 @@ void ControlConnection::reconnect(bool retry_current_host) {
 
   connection_ = new Connection(session_->loop(),
                                session_->config_,
+                               session_->metrics_.get(),
```
Comment (aholmberg) [inline src/control_connection.cpp]: Maybe use ``session->metrics()`` accessor?

File: src/session.cpp [hunk 3]
```diff
@@ -550,7 +586,7 @@ void Session::on_execute(uv_async_t* data) {
         Address address;
         if (!request_handler->get_current_host_address(&address)) {
           request_handler->on_error(CASS_ERROR_LIB_NO_HOSTS_AVAILABLE,
-                                    "Session: No hosts available");
+                                    "All connections on all I/O threads are busy");
```
Comment (aholmberg) [inline src/session.cpp]: Much more precise language. I like it.

File: src/metrics.hpp [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        id = reinterpret_cast<void*>(thread_id);
+        uv_key_set(&thread_id_key_, id);
+      }
+      return reinterpret_cast<size_t>(id) - 1;
+    }
+
+  private:
+    const size_t max_threads_;
+    boost::atomic<size_t> thread_count_;
+    uv_key_t thread_id_key_;
```
Comment (aholmberg) [inline src/metrics.hpp]: Are we abandoning libuv 0.10 on this version? This appears to be new in 1.4.2, 0.11.11.

Comment (aholmberg) [inline src/metrics.hpp]: https://github.com/libuv/libuv/commit/5d2434bf71e47802841bad218d521fa254d1ca2d

File: src/metrics.hpp [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+    , request_rates(&thread_state_)
+    , total_connections(&thread_state_)
+    , available_connections(&thread_state_)
+    , exceeded_pending_requests_water_mark(&thread_state_)
+    , exceeded_write_bytes_water_mark(&thread_state_)
+    , connection_timeouts(&thread_state_)
+    , pending_request_timeouts(&thread_state_)
+    , request_timeouts(&thread_state_) {}
+
+  ThreadState thread_state_;
```
Comment (aholmberg) [inline src/metrics.hpp]: Could be private. (shrug)

File: src/metrics.hpp [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+
+public:
+  Histogram requests;
```
Comment (aholmberg) [inline src/metrics.hpp]: "request_latency" or "request_latencies"?

File: src/request_handler.cpp [hunk 7]
```diff
@@ -90,7 +92,14 @@ bool RequestHandler::is_host_up(const Address& address) const {
...
 }
 
+void RequestHandler::start_request() {
+  start_time_ns_ = uv_hrtime();
+}
+
 void RequestHandler::set_response(Response* response) {
+  uint64_t elapsed = uv_hrtime() - start_time_ns_;
+  connection_->metrics()->requests.record_value(elapsed / 1000);
+  connection_->metrics()->request_rates.mark();
```
Comment (aholmberg) [inline src/request_handler.cpp]: Just a thought: would it make sense to to ``metrics()->record_request(<latency>)" and have the metrics class update the latency histo and mark the meter?

File: src/metrics.hpp [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+      free(histogram_);
+      uv_mutex_destroy(&mutex_);
+    }
+
+    void record_value(int64_t value) {
+      histograms_[thread_state_->current_thread_id()].record_value(value);
+    }
+
+    void get_snapshot(Snapshot* snapshot) const {
+      uv_mutex_lock(&mutex_);
```
Comment (aholmberg) [inline src/metrics.hpp]: ScopedMutex?

File: src/request_handler.cpp (hunk 7 above)
Comment (mpenick) [inline src/request_handler.cpp]: +1 good thought.

File: src/metrics.hpp (hunk 8 above)
Comment (mpenick) [inline src/metrics.hpp]: +1

File: src/metrics.hpp (hunk 6 above)
Comment (mpenick) [inline src/metrics.hpp]: Good idea.

File: src/metrics.hpp (hunk 5 above)
Comment (mpenick) [inline src/metrics.hpp]: It should be.

File: src/metrics.hpp (hunk 4 above)
Comment (mpenick) [inline src/metrics.hpp]: Argh!!!!! Good find.

**Comment by mpenick** [comment]
Final commit: 663351040af01ab5dfbfa05c60db0334c68a224d

**Pull request closed without merging**